[Feat/notification signals] Notification/ task,signls 작성

han-son-03 opened the pull request:
## ✅ PR 요약
- 작업 요약: 
     - signal, task 단위 테스트 코드 작성
     - Celery 기반 비동기 알림 처리 및 테스트 환경(eager mode)구성
## 📄 상세 내용
- [x] signals.py 
    -Django `post_save` signal을 사용하여 주요 이벤트 발생 시 `Notification` 모델이 생성
- [x] task.py
    -Celery task(`send_to_pubsub`)를 통해 외부 알림(Pub/Sub) 전송 처리
- [x] 테스트 코드 작성
  test_signal.py
    -이벤트 발생 시 `Notification`이 정상적으로 생성되는지 검증
    - Celery task가 호출되는지 `patch`를 통해 확인
  test_task.py
    - Celery task 실행 시 알림 전송 로직이 정상 동작하는지 검증
    - 외부 서비스 호출을 Mock 처리하여 단위 테스트로 구성
- [x] Celery eager 모드 설정
## ✏️주요 변경 사항
  1. Signal 기반 알림 생성 구조 도입
  2. Celery task 비동기 알림 처리 및 예외 대응 로직 추가
  3. 테스트 환경(eager mode) 기준의 안정적인 유닛 테스트 구성

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

han-son-03 committed:
✨ feat: 이벤트에 따른 실시간 알림 기능 작성

사용자가 특정 이벤트를 일으키면 post_save 될 때 알림이 지정된 그룹이나 사용자에게 알림이 발송 됩니다.

han-son-03 committed:
✨ feat: 일림 생성 비동기 처리(celery)

signal에서 이벤트를 감지하여 task를 넘기면 백그라운드에서 celery를 통해 브러커 전달 전 작업을 처리합니다.

han-son-03 committed:
💡 chore: celery 기본 설정 추가

celery에 필요한 기본적인 설정들을 추가 했습니다. base에 celery 필수 설정, __inti__.py 설정 추가, celery.py Celery beat 설정 포함 코드 작성

han-son-03 committed:
🎨 style: black, isort 코드 포매터 실행

포메터 실행

han-son-03 committed:
💡 chore: 불필요한 로그 제거 및 변수명 수정

todo 제거 및 swagger tag 통일화

han-son-03 committed:
✨ feat: 알림 백그라운드

알림 관련 백그라운드 작업에 필요한 파일 작성 했습니다 signal.py/ task.py + 관련 테스트 파일

han-son-03 committed:
🐛 fix: Celery Task 전송 과정에서 발생한 브로커 연결 오류 임시 해결

Django 테스트 환경에서 Celery Task 호출
Task가 메시지 브로커(Broker)인 Redis 연결시도 했으나 브로커가 실행되고 있지 않아 연결 거부
clery.pydp Eager Execution (즉시 실행) 모드를 활성화하여 임시 해결

han-son-03 committed:
✅ test: task test 코드 작성

다음과 같은 Notification 관련 Celery Task에 대한 테스트 코드를 추가했습니다.
테스트 환경은 Eager Mode를 사용하며, Task가 실행 시 Redis Pub/Sub으로
메시지를 정상적으로 발행하는지와, 비동기 리스너가 이를 성공적으로 수신하는지를 검증합니다.
- `send_to_pubsub`: 생성된 알림을 Redis Pub/Sub을 통해 실시간으로 전송하는 Task 테스트
- `send_study_group_notification`: 스터디 그룹 관련 알림을 전송하는 Task 테스트

han-son-03 committed:
🎨 style: black, isort 코드 포매터 실행

포메터 실행으로 공란 추가 및 삭제

han-son-03 committed:
💡 chore: 불필요한 로그 제거 및 변수명 수정

타입 힌트 관련 주석 삭제

han-son-03 committed:
🚚 build: Celery Task 테스트 환경 설정 (Eager Mode)

`config.settings.test.py` 파일에 Celery 테스트 환경 설정을 추가했습니다.
- `CELERY_TASK_ALWAYS_EAGER = True`, `CELERY_TASK_EAGER_PROPAGATES = True`
설정을 적용하여 테스트 코드 실행 시 Celery Task가 동기적으로 실행되도록 변경했습니다.
- 테스트 시 외부 종속성(Redis)을 제거하기 위해 `CELERY_BROKER_URL`을 `memory://`로,
`CELERY_RESULT_BACKEND`를 빈 문자열로 설정했습니다.

han-son-03 committed:
🎨 style: black, isort 코드 포매터 실행

black 실행 공란 생성

2joonkim commented on apps/notification/infra/task.py:
```diff
@@ -0,0 +1,135 @@
+                "type": notification.type,
+                "content": notification.content,
+                "back_ulr_link": notification.back_url_link,
+                "create_at": notification.created_at,
```
오타 수정해주세요!!

2joonkim commented on apps/notification/infra/task.py:
```diff
@@ -0,0 +1,135 @@
+        notifications = [
+            Notification(
+                user_id=participant.member.user_id.id,
+                content=f"rmadlf {participant.schedule.start_time.strftime('%H:%M')}부터"
```
```content=f"rmadlf ...```
이것도 오타인거 같은데 맞나용..?

2joonkim commented on apps/notification/infra/task.py:
한손님 이거 전체적으로 잘 작성하신 거 같은데 한줄한줄 뜯어보니
오타들이 조금씩 있네요 이거 한번만 디테일하게 살펴보시고 오타 수정해주세요!!

han-son-03 replied:
넵 금방 수정하겠습니다

2joonkim commented on config/settings/base.py:
```diff
@@ -236,3 +236,9 @@
+
+# celery 필수 설정
+CELERY_BROKER_URL = f"redis://{REDIS_HOST}:6379/1"  # 환경에 맞게 수정
+CELERY_RESULT_BACKEND = f"redis://{REDIS_HOST}:6379/2"  # 선택
```
이거 포트 하드코딩 하지 마시고 config/base 부분 보시면
REDIS_PORT = os.getenv("REDIS_PORT") 있어요 
하드코딩하지마시고 포트 변수 사용해주세요

``` f"redis://{REDIS_HOST}:{REDIS_PORT}/1" ```

2joonkim commented on apps/notification/tests/test_task.py:
```diff
@@ -0,0 +1,96 @@
+            try:
+                await listener_task
+            except asyncio.TimeoutError:
+                listener_task.cancel()
```
이쪽 테스트는 indent 잘못 쓰신거같은데
한번 확인해보시겠어요?
현재 group_message_listener 함수를 만드신 뒤
group_message_listener()를 호출하는 코드가 전혀 없어요

이렇게 되면 당연히 assertion이 없으니 터질일도 없고
빈값인데 당연히 의미없는 테스트만 하나 늘어나는게 되겠죵?

han-son-03 committed:
💡 chore: 오타 및 변수명 수정

task, signal 변수명 등 각각 오타수정 pubsub_creator 에러메시지 수정

han-son-03 committed:
✅ test: 기존 테스트 파일에서 발생하는 event loop에러로 인한 새로운 테스트 코드 작성

기존 테스트 코드는 비동기 로직이 동일한 테스트 환경에서 함께 실행되며 이벤트 루프 충돌
책임과 동작 의도를 검증하는 유닛 테스트로 변경 개인 알림, 스터디 그룹 알림, 일정 기반 알림 태스크를 각각 분리하여 테스트
테스트에서는 이벤트 루프를 직접 실행하지 않고 태스크 호출 여부만 검증하기 때문에 일부 코루틴이 await 되지 않은 상태로 종료

han-son-03 committed:
💡 chore: 불필요한 로그 제거 및 변수명 수정

주석 정리

han-son-03 committed:
🎨 style: black, isort 코드 포매터 실행

코드 포메터 실행

han-son-03 committed:
♻️ refactor: import 중복 코드 정리

Import 중복 정리

2joonkim commented on config/celery.py:
```diff
@@ -0,0 +1,22 @@
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "config.settings.local")
+
+app = Celery("config")
+app.config_from_object("django.conf:settings", namespace="CELERY")
+
+app.autodiscover_tasks()
+
+app.conf.beat_schedule = {
+    "send_tomorrow_schedule_notifications": {
+        "task": "apps.notification.infra.task.send_tomorrow_schedule_notifications",
+        "schedule": crontab(hour=0, minute=1),
+    },
+    "send_today_schedule_notifications": {
+        "task": "apps.notification.infra.task.send_today_schedule_notifications",
+        "schedule": crontab(hour=0, minute=1),
```
이 부분 전체경로 말고 name만 등록해주실래요?
beat에서 name만으로 등록되어 있어서 못 찾을겁니다

han-son-03 committed:
♻️ refactor: celery.py task 경로 간소화

`celery.py`에서 Task를 등록할 때 사용되던 경로를 전체 경로에서 Task 이름으로 수정

han-son-03 committed:
🐛 fix: conflicts문제 충돌 해결 및 통합

git conflicts문제 충돌 해결 및 통합

han-son-03 committed:
🎨 style: black, isort 코드 포매터 실행

black, isort 코드 포매터 실행

han-son-03 committed:
♻️ refactor: redis 포트 값 문제

redis 포트 값 문제 intf로

han-son-03 committed:
♻️ refactor: celery.py local-> base

celery.py local-> base

han-son-03 committed:
♻️ refactor: redis port 숫자로 하드코딩

redis port 숫자로 하드코딩

han-son-03 committed:
♻️ refactor: celery settings 오류 해결

celery settings 오류 해결
